gem/optimise.py

```diff
@@ -229,34 +255,101 @@ def delta_elimination(sum_indices, factors):
     return sum_indices, factors
 
 
-def associate_product(factors):
-    """Apply associativity rules to construct an operation-minimal product tree.
+ExpressionAndFlopCount = namedtuple('ExpressionAndFlopCount', ['expression', 'flops'])
+
 
-    For best performance give factors that have different set of free indices.
+def associate_operation(operands, type):
+    """Apply associativity rules to associative operations (e.g. summation and
+    product) to construct an operation-minimal operation tree.
+
+    :arg operands: list of operands
+    :arg type: GEM class which is associative, e.g. Sum, Product
+
+    :returns: named tuple ExpressionAndFlopCount, with expression holds the
+              resultant expression, and flops holds its number of flops
     """
-    if len(factors) > 32:
+    # group operands by their free indices
+    if type == Sum:
+        base = Zero()
+    elif type == Product:
+        base = one
+    else:
+        raise AssertionError(
+            "Unknown associative operation type {0}!".format(type))
```

Comment:
Be specific, or be generic, but in any case: be honest! You can have a function that does addition (specific), or a function that does multiplication (specific), or a function that works for any binary operator (generic), but what you did here something that looks like a generic function, but is really just two specific things squeezed into one and you can choose between the specific things by a function argument. If it really couldn't be generic, then it is better to have two functions with different names (honest).

In this concrete example you seem to break generality for absolutely no good reason, since the only thing you obtain by it (`base`) is something you do not even need.